### PR TITLE
Colon used as separator #83

### PR DESCRIPTION
**Pull request by kierun**
There is a new option called "`dividers`" that is a string of characters that could be use as separators as long as they are in the `string.printable` set.

There are two unit tests to make sure the functionality works. However, they are not exhaustive.

I think that the worst thing a user can do is pass some really weird characters in the above set that will cause the regex to return broken results. In that case, the parsing of the configuration file might fail or garbage could be loaded.

_Someone more versed in the code base and regex might want to have a look at the code!_

### CODE REVIEW

File: src/configobj/__init__.py [hunk 1]
```diff
@@ -1032,18 +1033,21 @@ class ConfigObj(Section):
...
+            (                       # keyword
+                (?:".*?")|          # double quotes
+                (?:'.*?')|          # single quotes
+                (?:[^'"{0}].*?)     # no quotes
+            )
+            \s*[{0}]\s*             # dividers
+            (.*)                    # value (including list values and comments)
+            $   # line end
+            '''.format(''.join(self._dividers)),
+            re.VERBOSE)
```
Comment (EliAndrewC) [inline src/configobj/__init__.py]: We should probably implement some caching here.  There's only one place that ``self._keyword`` is called, but it gets called a LOT.  By writing the property this way, we'll end up re-compiling the regex on almost every single line of the config file being parsed.  Admittedly, I haven't done any profiling, so it's possible that this is just something that seems bad but wouldn't matter in practice.

With that being said, it's an easy enough fix.  Instead of

```
return re.compile(...)
```

We can just say

```
if not hasattr(self, '__keyword'):
    self.__keyword = re.compile(...)
return self.__keyword
```

That will ensure that the regex only gets compiled the first time.

File: docs/configobj.rst [hunk 2]
```diff
@@ -444,6 +444,13 @@ ConfigObj takes the following arguments (with the default values shown) :
     creating a ConfigObj instance from a configspec file you must pass True
     for this argument as well as ``list_values=False``.
 
+* 'dividers': ``'='``
+
+    A string of character to use as dividers. It defaults to the equal sign
+    (`=`). For compatibility with `ConfigParser`, you can set it to
+    `diverders='=:'`. All characters not in `string.printable` will be
```
Comment (EliAndrewC) [inline docs/configobj.rst]: Minor typo: ``diverders`` instead of ``dividers``.

File: src/tests/test_configobj.py [hunk 3]
```diff
@@ -1375,3 +1375,28 @@ def test_writing_out_dict_value_with_unrepr(self):
...
+
+def test_dividers_unhappy(cfg_contents):
+    cfg = cfg_contents(u"""
+[section]
+    ook … 'eek'
+    """)
+    try:
+        c = ConfigObj(cfg, dividers=u'…')
+    except AttributeError as err:
+        assert str(err) == "No valide characters found for dividers."
```
Comment (EliAndrewC) [inline src/tests/test_configobj.py]: Great that you added unit tests!

There's one more test case I'd like to see, which is testing what happens when more than one divider appears on a line.  For example, I'd expect if someone set ``dividers=':='`` and then said

```
equation: p=np
```

then either ``equation`` will end up as ``'p=np'`` or the line will be treated as having more than one divider and will result in an error.  Either way we should have a test for this, since it's a new edge case.  (We should probably add a note in the docs you added explaining this as well.)

**Comment by EliAndrewC** [comment]
Great stuff, thanks for submitting this!

I had a few minor suggestions, but this looks ready to merge after the couple of tweaks I mentioned.

File: src/configobj/__init__.py (hunk 1 above)
Comment (kierun) [inline src/configobj/__init__.py]: This breaks a unit test: `TestMerging.test_merge_coupling[data1]` as you cannot deep copy the object then.

File: src/configobj/__init__.py [hunk 4]
```diff
@@ -1137,17 +1141,25 @@ def __init__(self, infile=None, options=None, configspec=None, encoding=None,
...
                     create_empty=False, file_error=False, stringify=True,
                     indent_type=None, default_encoding=None, unrepr=False,
-                    write_empty_values=False, _inspec=False)``
+                    write_empty_values=False, _inspec=False, dividers='=')``
         """
+
+        # Dividers, see issue #83.
+        self._dividers = [x for x in dividers if x in string.printable]
+        if not self._dividers:
+            raise AttributeError("No valide characters found for dividers.")
```
Comment (kierun) [inline src/configobj/__init__.py]: Fix, cursed dyslexia!

File: docs/configobj.rst (hunk 2 above)
Comment (kierun) [inline docs/configobj.rst]: Fix, cursed dyslexia!

File: src/tests/test_configobj.py (hunk 3 above)
Comment (kierun) [inline src/tests/test_configobj.py]: The *first* separator will be the one taken. Any other will just be part of the option. So, for example:

`Fred := some weird string`

Will yeild:

`assert c['section']['Fred'] == '= some weird string'`

Is this acceptable?

And yes, this should be documented too! ^_~

**Comment by kierun** [comment]
Couple of comments on the caching of the regexp (breaks deep copy) and a clarification of the multiple divider characters on the same line behaviour.

Thank you.

File: src/configobj/__init__.py (hunk 1 above)
Comment (kierun) [inline src/configobj/__init__.py]: `copy.copy` seems to do the trick.

**Comment by kierun** [comment]
Added some more changes which I hope answer all your (valid) criticism. Let me know if that's fine with you.

File: src/configobj/__init__.py [hunk 5]
```diff
@@ -757,7 +758,7 @@ def merge(self, indict, decoupled=False):
         """
         for key, val in list(indict.items()):
             if decoupled:
-                val = copy.deepcopy(val)
+                val = copy(val)
```
Comment (jhermann) [inline src/configobj/__init__.py]: There impact is quite unhidden, you defeat the purpose of the ``decoupled`` flag. Also problematic: changing this for a single regex is unwise, since (mostly) _all_ of them need to become dynamic for implementing in-line comments with alternate markers. A way better solution would be to put them all into a utility class (taking the necessary params to build the regex), that can then be aggregated and used by ConfigObj.

Comment (kierun) [inline src/configobj/__init__.py]: Just so I understand you: Take all the regexp into a utile class (with the delimiter parameter to `__init__()` and have ConfigObj instantiate one of those?

Also, I am unsure what you mean by all of them becoming dynamic. As far as I can see, only one (`_keyword`) uses the delimiter. Can you provide a test for what you mean?

Thinking about it, maybe '#' should not be allowed as a delimiter… Is there any other character that need to be excluded?

Comment (kierun) [inline src/configobj/__init__.py]: I tried to pull the `_keyword` regex out of `ConfigObj` and into a separate class. Unsurprisingly, the `deepcopy` still fails.  So, I must have misunderstood what you meant. My apologies, could you rephrase?

Comment (kierun) [inline src/configobj/__init__.py]: As a side note, the unit tests passed after I changed `deepcopy` to `copy.copy`. If that change is breaking, then there should be a unit test to reflect that. No blame, just trying to avoid anyone else falling into the same trap.

**Comment by kierun** [comment]
So, any idea as to how to proceed with this?

**Comment by kierun** [comment]
So, any idea as to how to proceed with this?

Clearly, the simplest way is to give up on dynamic setting and just add a `:` to the static regexp. It achieves compatibility but that's about it so I find it a little inelegant.

**Comment by kierun** [comment]
Anyone?…

**Comment by kierun** [comment]
It's been far too long… Clearly, no one (not even me!) cares enough.
